Passthrough "x-*" options to mtab

MatthewCash opened the pull request:
This implements #651, tested with bindfs.

There are 3 main changes to get this to work:

- Unknown options will be added to `mtab_opts` if they start with "x-"
- Options starting with "x-" are not provided to `do_mount()`
- Options starting with "x-" are added back for the call to `add_mount()`

On my system, `/etc/mtab` is a symlink to `/proc/self/mounts`. This causes fuse to skip updating the mtab entry because of the symlink check. I removed the check in this patch, but there is likely a reason it was there in the first place, maybe the check could be narrowed to links that are pointing to `/proc/self/mounts`?

Also, another set of eyes to make sure my string manipulation logic is correct would be appreciated.

bsbernd commented on lib/mount.c:
```diff
@@ -209,6 +209,11 @@ static int fuse_mount_opt_proc(void *data, const char *arg, int key,
 	case KEY_MTAB_OPT:
 		return fuse_opt_add_opt(&mo->mtab_opts, arg);
+
+	case FUSE_OPT_KEY_OPT:
```
Could you please add a short comment here explaining what it is for? Like

/* Third party options like 'x-gvfs-notrash' */

MatthewCash replied:
Done

bsbernd reviewed (changes requested):
Thank you, sounds very useful.

I'm going to review the option parsing later on, but could please also change the commit message? I don't mind referencing an issue, but the commit still should be self explaining. I read code quite often without a browser...

In the commit message like

This implements #651, tested with bindfs. Basically #651 is about some 3rd party mount options like 'x-gvfs-notrash'.

bsbernd commented:
Hmm, tests fail because of strlcpy. Looks like recent gnu libc has strlcpy, but then we also have users with other libcs like uClibc.

MatthewCash commented:
> I'm going to review the option parsing later on, but could please also change the commit message? I don't mind referencing an issue, but the commit still should be self explaining. I read code quite often without a browser...

Done

> Hmm, tests fail because of strlcpy. Looks like recent gnu libc has strlcpy, but then we also have users with other libcs like uClibc.

I replaced them all with the "strn*" versions.

ashleypittman commented:
> On my system, `/etc/mtab` is a symlink to `/proc/self/mounts`. This causes fuse to skip updating the mtab entry because of the symlink check. I removed the check in this patch, but there is likely a reason it was there in the first place, maybe the check could be narrowed to links that are pointing to `/proc/self/mounts`?

Was there a problem with the check?  It sounds like in your case it was doing it's job correctly?

MatthewCash commented:
> Was there a problem with the check? It sounds like in your case it was doing it's job correctly?

What is the purpose of the check? It prevents fusermount from ever updating mtab on my system, is this intended behavior?

ashleypittman commented:
> > Was there a problem with the check? It sounds like in your case it was doing it's job correctly?
> 
> What is the purpose of the check? It prevents fusermount from ever updating mtab on my system, is this intended behavior?

If mtab is a symlink to `proc/self/mounts` then it doesn't need updating from userspace as the kernel maintains it I thought?  Where is a symlink to in your system?

MatthewCash commented:
> > > Was there a problem with the check? It sounds like in your case it was doing it's job correctly?
> > 
> > 
> > What is the purpose of the check? It prevents fusermount from ever updating mtab on my system, is this intended behavior?
> 
> If mtab is a symlink to `proc/self/mounts` then it doesn't need updating from userspace as the kernel maintains it I thought? Where is a symlink to in your system?

I see, the issues is that the kernel is not passed the x-* options so the kernel-maintained mtab is not correct.

For this change to work, fuse would always need to update mtab if there are x-* options. Are there any downsides to always updating mtab (given that it is possible)?

MatthewCash commented:
> > > > > Was there a problem with the check? It sounds like in your case it was doing it's job correctly?
> > > > 
> > > > 
> > > > What is the purpose of the check? It prevents fusermount from ever updating mtab on my system, is this intended behavior?
> > > 
> > > 
> > > If mtab is a symlink to `proc/self/mounts` then it doesn't need updating from userspace as the kernel maintains it I thought? Where is a symlink to in your system?
> > 
> > 
> > I see, the issues is that the kernel is not passed the x-* options so the kernel-maintained mtab is not correct.
> > For this change to work, fuse would always need to update mtab if there are x-* options. Are there any downsides to always updating mtab (given that it is possible)?
> 
> I'll have to look closer at what you're doing because I don't fully understand it but I'd expect it to fail if you tried to modify /proc/self/mounts and, if you tried to replace the symlink with a file then this would fail for all but the root user and if you were running as root and replaced the symlink with a file then it would then be wrong for all other processes.
> 
> One oddity with mounts is that different processes can see different filesystems mounted, which is why it's a symlink in the first place, if mtab is ever a regular file and modified from userspace then I'd expect the system to be a minimal boot image or similar.

libfuse/fusermount isn't directly touching any tab files. It calls `/bin/mount` instead. I can manually run `mount --no-canonicalize -i -f -o x-test ...` to accomplish the same behavior.

MatthewCash commented:
After stracing the `/bin/mount` command, it appears that `/run/mount/utab` is the file that is updated with the new options, which somehow (maybe the kernel does it?) causes `/proc/self/mounts` to reflect them.

bsbernd commented on lib/mount_util.c:
```diff
@@ -65,9 +64,6 @@ static int mtab_needs_update(const char *mnt)
 
-		if (S_ISLNK(stbuf.st_mode))
-			return 0;
-
```
Ok, I see this  is what the symlink discussion is about - we first need to check if there would be side effects.

MatthewCash replied:
I wonder if any of these mtab checks are actually necessary, I would assume that libmount would not try to edit the tab if its nonexistent/unwritable?

bsbernd replied:
I really don't know. So this line comes from commit 5f28cd15ab43c741f6d116be4d3a9aa5d82ab385, which explains what the code does, but not why...

```
commit 5f28cd15ab43c741f6d116be4d3a9aa5d82ab385
Author: Miklos Szeredi <miklos@szeredi.hu>
Date:   Thu Jul 10 19:35:21 2008 +0000

    Skip calling mount(8) if /etc/mtab doesn't exist...
```

Exactly the reason why I typically insist on more verbose commit messages. Well, the commit also has a ChangeLog entry

```
diff --git a/ChangeLog b/ChangeLog
index 638dd3cb..4f565668 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -3,6 +3,10 @@
        * Fix handling of (no)suid and (no)dev options if filesystem is
        mounted from /etc/fstab or via mount(8).  Reported by Jan Ondrej.
 
+       * Skip calling mount(8) if /etc/mtab doesn't exist or if it's on a
+       read-only filesystem.  This works around issues with certain mount
+       implementations.  Reported by Szabolcs Szakacsits.
+
```

No idea if these issues do still exist - I think any recent distribution has mtab as symlink...

bsbernd replied:
I actually start to wonder if libfuse should link to libmount, because there are functions available that let you check for a writable utab. But certainly not in this commit, just something we might consider in the future.

bsbernd commented:
Before merging we first need to check side effects of the symlink check.

Could you please also use max 80 char per line, better 69 char,  in the commit message, please?

MatthewCash commented:
> Before merging we first need to check side effects of the symlink check.
> 
> Could you please also use max 80 char per line, better 69 char, in the commit message, please?

Is there anything specific that you would like me to test? And do you mean that commit message lines should be limited to 69 chars?

bsbernd commented:
Yes please, max 80 char per line in the commit message. I wouldn't like to put in a hard limit, in case one needs to paste links or back traces. Similar for code, lines should be in between 80 and 100 chars - comes from kernel coding limits.

bsbernd commented:
@MatthewCash I don't know something specific we can test for, right now I'm grepping through util-linux for utab.  For example interesting is

```    /*
     * libmount uses utab.event file to monitor and control utab updates
     */
```

Do know if gnome is linked to libmount? I'm using XFCE and don't have all the gnome applications installed...

MatthewCash commented:
> @MatthewCash I don't know something specific we can test for, right now I'm grepping through util-linux for utab. For example interesting is
> 
> ```
>      * libmount uses utab.event file to monitor and control utab updates
>      */
> ```
> 
> Do know if gnome is linked to libmount? I'm using XFCE and don't have all the gnome applications installed...

It's hard to tell because libmount is usually (always?) packaged in util-linux, but it appears that [glib uses it on NixOS](https://github.com/NixOS/nixpkgs/blob/8b1ad909bdda5819a7c71c89d9bffea029a67a8e/pkgs/development/libraries/glib/default.nix#L113).

bsbernd commented:
I guess we should be ok

```
util-linux 2.39 Release Notes
=============================

Release highlights
------------------

mount(8) and libmount:


 * no more supports --enable-libmount-support-mtab. Yes, classic
   writable /etc/mtab is completely dead now.
```


@MatthewCash Could you please add in the commit message that the symlink check for /etc/mtab is removed? Something like

```
This also removes the test if /etc/mtab is a symlink. This test was added in commit 5f28cd15ab43c741f6d116be4d3a9aa5d82ab385 and the corresponding ChangeLog entry in this commit points to mount issues for read-only mtab.
However, in all recent Linux distributions /etc/mtab is a symlink to /proc/self/mounts and never writable. In fact, util-linux 2.39 (libmount) entirely removed support for a writable mtab. At least since util-linux 2.19 (10-Feb-2011) /run/mount/utab is used as replacement for userspace mount entries..
```

MatthewCash commented on util/fusermount.c:
```diff
@@ -1170,7 +1202,18 @@ static int mount_fuse(const char *mnt, const char *opts, const char **type)
-		res = add_mount(source, mnt, *type, mnt_opts);
+		// Add back the options starting with "x-"
+		add_mount_opts_size = strlen(mnt_opts) + strlen(x_opts) + 2;
+		add_mount_opts = malloc(add_mount_opts_size);
```
The original implementation does not use `opts` for both `do_mount()` and `add_mount()`, the options for `add_mount` are stored in `mnt_opts` by the `do_mount` call.

I'll edit the comment, something like "Add back the options starting with "x-" to opts from do_mount" ?

bsbernd replied:
Hmm, ok, question is 'why is it not using all options'. I'm ok with it for with this commit, but I keep the freedom to check why it is done that way and then 

a) to change to all options 
or 
b) to add a comment why it is done that way

I also don't know all code here and it is confusing for me without a comment explaining the reason.

MatthewCash replied:
Yeah, I'm not sure either I just assumed is was like that for a reason. 🤷

MatthewCash replied:
It looks like `do_mount()` adds some more (default?) options to the list like `rw` and `userid=...`, I don't think it every removes options but I'm not sure.

bsbernd replied:
> Yeah, I'm not sure either I just assumed is was like that for a reason. 🤷

Well, this is exactly what I mean - nobody knows why, except maybe the original author who probably has forgotten as well in the mean time.

>It looks like do_mount() adds some more (default?) options to the list like rw and userid=..., I don't >think it every removes options but I'm not sure.

I'm going to test myself what would happen with full options.

MatthewCash commented on util/fusermount.c:
```diff
@@ -97,6 +97,27 @@ static struct mntent *GETMNTENT(FILE *stream)
+/*
+ * Take a ',' separated option string and extract "x-" options
+ */
+static void split_opts(const char *original, char *non_x_opts, char *x_opts)
```
Done

MatthewCash commented on util/fusermount.c:
```diff
@@ -97,6 +97,27 @@ static struct mntent *GETMNTENT(FILE *stream)
+
+	for (opt = original; opt < original + len; opt = opt_end + 1) {
+		opt_end = memchr(opt, ',', len);
+		if (opt_end == NULL) opt_end = original + len;
```
I updated the conditions formatting. I took a look at the [kernel coding style guide](https://www.kernel.org/doc/html/next/process/coding-style.html) but it is very surface level so I might be missing some other style changes.

bsbernd commented:
Merged, @MatthewCash thank you for your contribution!

MatthewCash commented:
Great, thank you for all your help!

bsbernd commented:
Just noticed that we have a bit of testing issue when it comes to fusermount changes - need to check for the reason, but I think it does not install fusermount3 but uses system default. So after installing fusermount3 and running it with ASAN

```
=================================================================
==7131==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x607000000147 at pc 0x7f6357a3cb11 bp 0x7ffdea8dc8b0 sp 0x7ffdea8dc060
READ of size 72 at 0x607000000147 thread T0
    #0 0x7f6357a3cb10 in __interceptor_strlen ../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:370
    #1 0x56283c215963 in extract_x_options ../util/fusermount.c:118
    #2 0x56283c215963 in mount_fuse ../util/fusermount.c:1187
    #3 0x56283c215963 in main ../util/fusermount.c:1517
    #4 0x7f6356e4fd09 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x23d09)
    #5 0x56283c21b9e9 in _start (/usr/local/bin/fusermount3+0x189e9)

0x607000000147 is located 0 bytes to the right of 71-byte region [0x607000000100,0x607000000147)
allocated by thread T0 here:
    #0 0x7f6357aa9e8f in __interceptor_malloc ../../../../src/libsanitizer/asan/asan_malloc_linux.cpp:145
    #1 0x56283c2158ce in mount_fuse ../util/fusermount.c:1185
    #2 0x56283c2158ce in main ../util/fusermount.c:1517
    #3 0x7f6356e4fd09 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x23d09)

SUMMARY: AddressSanitizer: heap-buffer-overflow ../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:370 in __interceptor_strlen
Shadow bytes around the buggy address:
  0x0c0e7fff7fd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0e7fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0e7fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0e7fff8000: fa fa fa fa fd fd fd fd fd fd fd fd fd fa fa fa
  0x0c0e7fff8010: fa fa 00 00 00 00 00 00 00 00 00 01 fa fa fa fa
=>0x0c0e7fff8020: 00 00 00 00 00 00 00 00[07]fa fa fa fa fa 00 00
  0x0c0e7fff8030: 00 00 00 00 00 00 07 fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8040: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8050: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8060: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e7fff8070: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==7131==ABORTING

```

MatthewCash commented:
> Just noticed that we have a bit of testing issue when it comes to fusermount changes - need to check for the reason, but I think it does not install fusermount3 but uses system default. So after installing fusermount3 and running it with ASAN

I'm not very familiar with Github Actions, but could we do something like `install util/fusermount3 /bin/` between lines 50 and 51 in `ci-build.sh`?

bsbernd commented:
I'm just fixing the issue (and changing the code a bit), will then look into ci-build.sh.

bsbernd commented:
@MatthewCash We found another issue (just in the process of enabling cppcheck for libfuse at my company (DDN)) - this time in my updated code: https://github.com/libfuse/libfuse/pull/924
